DifferentialDrive.md

```diff
@@ -0,0 +1,545 @@
+##### MAKE SURE TO CHECK OUT THIS [GUIDE](https://docs.wpilib.org/en/stable/docs/software/wpilib-tools/robot-simulation/unit-testing.html) FROM WPILIB ON HOW TO SETUP UNIT TESTS.
+
+Let’s talk about the tests for our drivetrain. We'll start it by creating a driveTest.java file in the robot folder of the tests folder. 
+
+First, we need to set things up before we run any tests. This is done in the setup() method:
+```java
+Drive drive;
+CANSparkMax spark;
+
+  @BeforeEach
+  public void setup() {
+    setupTests();
+    drive = new Drive();
+    spark = drive.getSpark();
+  }
+```
+- ```getSpark()``` is a simple method in drive.java that lets us use one of the motors: 
+    - ``` public CANSparkMax getSpark(){return leftLeader;}```
+
+Next, we will need to reset our drive between every tests so we are able to experiment different things at once. 
+```java
+  @AfterEach
+  void destroy() throws Exception {
+    reset(drive);
+  }
+```
+
+Once everything is set up, you can make tests with the `@Test` annotation. Here's a quick example of testing if our motors are getting voltage:
+```java
+   @Test
+  public void testSparks() {
+    drive.drive(1, 0);
+
+    assert spark.getBusVoltage() > 3;
+  }
+  ```
+  - 3 is a random number here, but it is small on purpose. The motor will need more voltage than 3 in order to properly run.
+
+  Another example: Here we are testing to see if our pose is changing as we drive: 
+```java
+  @Test
+  public void testPose(){
+    drive.drive(0.8, 0.8);
+    fastForward();
+
+    assertNotEquals(new Pose2d(), drive.getPose());
+  }
+```
+  #### Running the tests
+To actually see if the tests are valid, open the wpilib command palette(ctrl + shift + p) and search ```Test Robot Code``` and press enter. After a bit, it will show you the results of all your tests in the terminal. 
+
+![alt text](image-3.png)
+
+Unit tests should be made for most of your subsystems and are a great way to see if the logic behind the code actually makes sense and works. You should also try to test all parts of the subsystems to cover everything. 
+
+Here's another simple test but for LEDs. 
+```
+@Test
+  public void testRainbowTheme() {
+    var rainbow1 = LedStrip.getBufferDataString(led.rainbowAddressableLEDBuffer());
+    fastForward(2);
+    var rainbow2 = LedStrip.getBufferDataString(led.rainbowAddressableLEDBuffer());
+    assertNotEquals(rainbow1, rainbow2);
+  }
+  ```
+  If you just scan through the code, its easy to tell what the process is and what we're testing. Setting up the leds to a pattern, forwarding in time so that its a different color than before, and testing to make sure that the two colors are indeed different from each other. 
```

Comment:
```suggestion
If you just scan through the code, it's easy to tell what the process is and what we're testing. We are setting up the LEDs to display a rainbow pattern, fast forwarding in time so that it's a different color than before, and testing to make sure that the two colors are indeed different from each other. 
```